central/notifiers/jira/jira.go

```diff
@@ -69,57 +68,182 @@ type jira struct {
 
-func getIssueTypes(client *jiraLib.Client, project string) ([]*jiraLib.MetaIssueType, error) {
-	// Low level HTTP client call is used here due to the deprecation/removal of the Jira endpoint used by the Jira library
-	// to fetch the CreateMeta data, and there is no API call for the suggested endpoint to use in place of the removed one.
-	// Info here:
-	// https://confluence.atlassian.com/jiracore/createmeta-rest-endpoint-to-be-removed-975040986.html
-	urlPath := fmt.Sprintf("rest/api/2/issue/createmeta/%s/issuetypes", project)
+type issueField struct {
+	Name    string `json:"name"`
+	Key     string `json:"key"`
+	FieldId string `json:"fieldId"`
+}
+
+type issueFieldsResult struct {
+	StartAt          int           `json:"startAt"`
+	MaxResults       int           `json:"maxResults"`
+	Total            int           `json:"total"`
+	IssueFields      []*issueField `json:"values"`
+	IssueFieldsCloud []*issueField `json:"fields"`
+}
+
+type permissionResult struct {
+	Permissions map[string]struct {
+		HavePermission bool
+	}
+}
+
+type jiraField struct {
+	Name string
+}
+
+func callJira(client *jiraLib.Client, urlPath string, result interface{}, startAt int) error {
+	if strings.Contains(urlPath, "?") {
+		urlPath = urlPath + fmt.Sprintf("&startAt=%d", startAt)
+	} else {
+		urlPath = urlPath + fmt.Sprintf("?startAt=%d", startAt)
+	}
 
 	req, err := client.NewRequest("GET", urlPath, nil)
 	if err != nil {
-		return []*jiraLib.MetaIssueType{}, err
+		return err
 	}
 
 	resp, err := client.Do(req, nil)
 	if err != nil {
-		return []*jiraLib.MetaIssueType{}, err
+		return err
 	}
 
-	result := &issueTypeResult{}
 	defer utils.IgnoreError(resp.Body.Close)
 	err = json.NewDecoder(resp.Body).Decode(result)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getIssueTypes(client *jiraLib.Client, project string) ([]*jiraLib.MetaIssueType, error) {
+	urlPath := fmt.Sprintf("rest/api/2/issue/createmeta/%s/issuetypes", project)
+
+	result := issueTypeResult{}
+
+	err := callJira(client, urlPath, &result, 0)
+
 	if err != nil {
 		return []*jiraLib.MetaIssueType{}, err
 	}
 
-	return result.IssueTypes, nil
+	returnList := make([]*jiraLib.MetaIssueType, result.Total)
+
+	if len(result.IssueTypes) == 0 {
+		returnList = result.IssueTypesCloud
+	} else {
+		returnList = result.IssueTypes
+	}
+
+	for result.Total != len(returnList) {
+		result = issueTypeResult{}
+		callJira(client, urlPath, &result, len(returnList))
+
+		if err != nil {
+			return []*jiraLib.MetaIssueType{}, err
+		}
+
+		var actualIssueTypes []*jiraLib.MetaIssueType
+		if len(result.IssueTypes) == 0 {
+			actualIssueTypes = result.IssueTypesCloud
+		} else {
+			actualIssueTypes = result.IssueTypes
+		}
+
+		returnList = append(returnList, actualIssueTypes...)
+	}
+
+	return returnList, nil
 }
 
-func isPriorityNeeded(client *jiraLib.Client, project, issueType string) (bool, error) {
+func getIssueFields(client *jiraLib.Client, project, issueId string) ([]*issueField, error) {
```

Comment:
I thought about this, too.  I landed on a hybrid approach where I made `callJira` generic but kept the "outer" functions largely duplicated, but I'll reconsider making them more generic as well.

Reply:
I thought about it some, and I think for now I'm just going to leave the duplication. I think it can be done using some combination of generics and interfaces, but perhaps I'll leave the thought experiment for another time :)